streams/src/main/java/org/apache/kafka/streams/kstream/internals/KTableSource.java

```diff
@@ -119,8 +134,13 @@ public void process(final K key, final V value) {
                 } else {
                     oldValue = null;
                 }
-                store.put(key, ValueAndTimestamp.make(value, context().timestamp()));
-                tupleForwarder.maybeForward(key, value, oldValue);
+                final boolean isDifferentValue = 
+                    store.putIfDifferentValues(key, ValueAndTimestamp.make(value, context().timestamp()), tuple.serializedValue);
```

Comment:
Q: If the record is out-of-order, shouldn't you put the value into the state store regardless of whether it is different from the old value or not? I think this would be more correct because the old value is actually the idempotent update to the new out-of-order value. This is one of the rare cases where we can correct the order. WDYT?

Reply:
Thanks @cadonna. I'd like to ask for clarification:

Do you mean, given the following sequence of input records:
```
A: 2 @time3
A: 2 @time4
A: 2 @time1
```

We would take the following actions:
```
put(A, 2 @time3) and forward the update
skip idempotent update (A is still 2 @time3)
put(A, 2 @time1) and forward the update (so now A is still 2, but the timestamp is updated to 1)
```

This might be controversial. I believe that when we "complete" the implementation of timestamp semantics in Streams, the current proposal is to actually drop disordered updates. In that context, it might seem like a step in the wrong direction to specifically implement the code to preserve them when we would otherwise drop them.

On the other hand, the _current_ semantics do seem to be better maintained if we take your suggestion. Perhaps we should let the future take care of itself and implement correct semantics as we currently understand them. I.e., I'm +1 on your suggestion.

Reply:
Yes, this is what I mean and the current semantics were my motivation to ask the question.

Reply:
These are good points! However, if this is the case, then we probably need to compare timestamps as well (i.e. to check if one is smaller than the other). We just need to modify the comparator in the serializer.

Reply:
Also, added some changes to accomodate this in the test.